feat: introduce OneShare subplugin

santhoshh-kumar opened the pull request:
<!-- Thanks for submitting a Pull Request to kestra. To help us review your contribution, please follow the guidelines below:

- Make sure that your commits follow the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) specification e.g. `feat(ui): add a new navigation menu item` or `fix(core): fix a bug in the core model` or `docs: update the README.md`. This will help us automatically generate the changelog.
- The title should briefly summarize the proposed changes.
- Provide a short overview of the change and the value it adds.
- Share a flow example to help the reviewer understand and QA the change.
- Use "close" to automatically close an issue. For example, `close #1234` will close issue #1234. -->

### What changes are being made and why?

Closes kestra-io/kestra#11572

This PR introduces a comprehensive **OneShare** subplugin for Microsoft 365, enabling seamless integration with OneDrive and SharePoint file storage services. The subplugin provides a complete suite of file management tasks and a real-time trigger for monitoring file changes.

#### Features Added

**Tasks:**
- **`Upload`**: Upload files to OneDrive with support for both small files (direct upload) and large files (resumable upload sessions). Includes configurable chunking, retry logic, and conflict resolution.
- **`Download`**: Download files from OneDrive to Kestra's internal storage.
- **`ListFiles`**: List all files and folders within a specified directory with full metadata (name, size, MIME type, timestamps, etc.).
- **`Create`**: Create new files or folders in OneDrive.
- **`Delete`**: Delete files or folders from OneDrive.
- **`Export`**: Export files to different formats (e.g., convert Word docs to PDF).

**Trigger:**
- **`Trigger`**: Real-time monitoring using Microsoft Graph Delta API to detect newly created files in a specified folder path. Supports polling intervals and maintains state across executions.

**Models:**
- **`OneShareFile`**: A unified data model representing files/folders with essential metadata (ID, name, MIME type, size, timestamps, web URL, folder flag).

**Infrastructure:**
- **`AbstractOneShareTask`**: Base class for all OneShare tasks extending Microsoft 365 Graph Connection.
- **`package-info.java`**: Proper plugin metadata with categories (STORAGE, CLOUD).
- Comprehensive unit tests for all tasks and models with proper mocking.
- Integration test flows in `src/test/resources/flows/oneshare/`.
- Test utilities (`OnesShareTestUtils`) for consistent test setup.

---

#### Example Flows for QA

``` yaml
id: oneshare-complete-demo
namespace: io.kestra.tests

inputs:
  - id: file
    type: FILE
    description: File to upload for testing

tasks:
  # Create a test folder
  - id: create_folder
    type: io.kestra.plugin.microsoft365.oneshare.Create
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    parentId: "root"
    name: "kestra-test-folder"
    itemType: FOLDER

  # Upload a file to the created folder
  - id: upload_file
    type: io.kestra.plugin.microsoft365.oneshare.Upload
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    parentId: "{{ outputs.create_folder.file.id }}"
    fileName: "uploaded-test-file.txt"
    from: "{{ inputs.file }}"
    conflictBehavior: REPLACE
    largeFileThreshold: 4194304
    maxSliceSize: 3145728

  # Create a text file with content in the folder
  - id: create_text_file
    type: io.kestra.plugin.microsoft365.oneshare.Create
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    parentId: "{{ outputs.create_folder.file.id }}"
    name: "created-file.txt"
    content: "This is test content created by Kestra"

  # List all files in the folder
  - id: list_files
    type: io.kestra.plugin.microsoft365.oneshare.ListFiles
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    itemId: "{{ outputs.create_folder.file.id }}"

  # Download the uploaded file
  - id: download_file
    type: io.kestra.plugin.microsoft365.oneshare.Download
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    itemId: "{{ outputs.upload_file.file.id }}"

  # Delete the created text file
  - id: delete_text_file
    type: io.kestra.plugin.microsoft365.oneshare.Delete
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    itemId: "{{ outputs.create_text_file.file.id }}"

  # Delete the uploaded file
  - id: delete_uploaded_file
    type: io.kestra.plugin.microsoft365.oneshare.Delete
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    itemId: "{{ outputs.upload_file.file.id }}"

  # Delete the test folder
  - id: delete_folder
    type: io.kestra.plugin.microsoft365.oneshare.Delete
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    itemId: "{{ outputs.create_folder.file.id }}"

  # Log summary
  - id: log_summary
    type: io.kestra.plugin.core.log.Log
    message: |
      OneShare Demo Completed Successfully!
      - Created folder: {{ outputs.create_folder.file.name }} (ID: {{ outputs.create_folder.file.id }})
      - Uploaded file: {{ outputs.upload_file.file.name }} (Size: {{ outputs.upload_file.file.size }} bytes)
      - Created text file: {{ outputs.create_text_file.file.name }}
      - Listed {{ outputs.list_files.count }} files in folder
      - Downloaded file to: {{ outputs.download_file.uri }}
      - All cleanup completed
```

``` yaml
id: oneshare-export-demo
namespace: io.kestra.tests

tasks:
  - id: create_doc
    type: io.kestra.plugin.microsoft365.oneshare.Create
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    parentId: "root"
    name: "test-export.docx"
    content: "Test document for PDF export"

  - id: export_to_pdf
    type: io.kestra.plugin.microsoft365.oneshare.Export
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    itemId: "{{ outputs.create_doc.file.id }}"
    format: PDF

  - id: cleanup
    type: io.kestra.plugin.microsoft365.oneshare.Delete
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    itemId: "{{ outputs.create_doc.file.id }}"
```

``` yaml
id: oneshare-trigger-demo
namespace: io.kestra.tests

triggers:
  - id: monitor_new_files
    type: io.kestra.plugin.microsoft365.oneshare.Trigger
    tenantId: "{{ globals.tenantId }}"
    clientId: "{{ globals.clientId }}"
    clientSecret: "{{ globals.clientSecret }}"
    driveId: "{{ globals.driveId }}"
    path: "/"
    interval: PT1M
    on: CREATE

tasks:
  - id: log_new_files
    type: io.kestra.plugin.core.log.Log
    message: |
      New files detected: {{ trigger.count }}
      Files: {{ trigger.files | jq('map(.name) | join(", ")') }}

  - id: process_each_file
    type: io.kestra.plugin.core.flow.ForEach
    values: "{{ trigger.files }}"
    tasks:
      - id: download_triggered_file
        type: io.kestra.plugin.microsoft365.oneshare.Download
        tenantId: "{{ globals.tenantId }}"
        clientId: "{{ globals.clientId }}"
        clientSecret: "{{ globals.clientSecret }}"
        driveId: "{{ globals.driveId }}"
        itemId: "{{ taskrun.value.id }}"

      - id: log_processed_file
        type: io.kestra.plugin.core.log.Log
        message: "Processed file: {{ taskrun.value.name }} ({{ taskrun.value.size }} bytes)"
```
      

---

### Setup Instructions

To use the OneShare subplugin, you need to configure Azure AD authentication:

#### Azure AD App Registration

1. **Create App Registration**:
   - Navigate to [Azure Portal](https://portal.azure.com/) → **Azure Active Directory** → **App registrations**
   - Click **New registration**
   - Provide a name (e.g., "Kestra OneShare Integration")
   - Select **Accounts in this organizational directory only**
   - Click **Register**

2. **Note Credentials**:
   - Copy the **Application (client) ID**
   - Copy the **Directory (tenant) ID**

3. **Create Client Secret**:
   - Go to **Certificates & secrets** → **Client secrets**
   - Click **New client secret**
   - Provide a description and select expiration
   - **Copy the secret value immediately** (won't be shown again)

4. **Grant API Permissions**:
   - Go to **API permissions** → **Add a permission**
   - Select **Microsoft Graph** → **Application permissions**
   - Add the following permissions:
     - `Files.ReadWrite.All` - Read and write files in all site collections
     - `Sites.ReadWrite.All` - Read and write items in all site collections
   - Click **Grant admin consent for your organization**

5. **Get Drive ID**:
   - For OneDrive: Use [Graph Explorer](https://developer.microsoft.com/en-us/graph/graph-explorer) and query `/me/drive` to get your drive ID
   - For SharePoint: Query `/sites/{site-id}/drive` or `/sites/{site-id}/drives`

---

### Contributor Checklist ✅

- [x] PR Title and commits follows [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Add a `closes #ISSUE_ID` or `fixes #ISSUE_ID` in the description if the PR relates to an opened issue.
- [x] Documentation updated (plugin docs from `@Schema` for properties and outputs, `@Plugin` with examples, `README.md` file with basic knowledge and specifics).
- [x] Setup instructions included if needed (API keys, accounts, etc.).
- [x] Prefix all rendered properties by `r` not `rendered` (eg: `rHost`).
- [x] Use `runContext.logger()` to log enough important infos where it's needed and with the best level (DEBUG, INFO, WARN or ERROR).

⚙️ **Properties**
- [x] Properties are declared with `Property<T>` carrier type, do **not** use `@PluginProperty`.
- [x] Mandatory properties must be annotated with `@NotNull` and checked during the rendering.
- [x] You can model a JSON thanks to a simple `Property<Map<String, Object>>`.

🌐 **HTTP**
- [x] Must use Kestra's internal HTTP client from `io.kestra.core.http.client` *(N/A - using Microsoft Graph SDK)*

📦 **JSON**
- [x] If you are serializing response from an external API, you may have to add a `@JsonIgnoreProperties(ignoreUnknown = true)` at the mapped class level. So that we will avoid to crash the plugin if the provider add a new field suddenly.
- [x] Must use Jackson mappers provided by core (`io.kestra.core.serializers`)

✨ **New plugins / subplugins**
- [x] Make sure your new plugin is configured like mentioned [here](https://kestra.io/docs/plugin-developer-guide/gradle#mandatory-configuration).
- [x] Add a `package-info.java` under each sub package respecting [this format](https://github.com/kestra-io/plugin-odoo/blob/main/src/main/java/io/kestra/plugin/odoo/package-info.java) and choosing the right category.
- [x] Icons added in `src/main/resources/icons` in SVG format and not in thumbnail (keep it big):
  - `plugin-icon.svg`
  - One icon per package, e.g. `io.kestra.plugin.aws.svg`
  - For subpackages, e.g. `io.kestra.plugin.aws.s3`, add `io.kestra.plugin.aws.s3.svg`
    See example [here](https://github.com/kestra-io/plugin-elasticsearch/blob/master/src/main/java/io/kestra/plugin/elasticsearch/Search.java#L76).
- [x] Use `"{{ secret('YOUR_SECRET') }}"` in the examples for sensible infos such as an API KEY.
- [x] If you are fetching data (one, many or too many), you must add a `Property<FetchType> fetchType` to be able to use `FETCH_ONE`, `FETCH` and even `STORE` to store big amount of data in the internal storage. *(N/A - tasks don't fetch large data sets)*
- [x] Align the `"""` to close examples blocks with the flow id.

🧪 **Tests**
- [x] Unit Tests added or updated to cover the change (using the `RunContext` to actually run tasks).
- [x] Add sanity checks if possible with a YAML flow inside `src/test/resources/flows`.
- [x] Avoid disabling tests for CI. Instead, configure a local environment whenever it's possible with `.github/setup-unit.sh` (which can be executed locally and in the CI) all along with a new `docker-compose-ci.yml` file (do **not** edit the existing `docker-compose.yml`).
- [ ] Provide screenshots from your QA / tests locally in the PR description. The goal here is to use the JAR of the plugin and directly test it locally in Kestra UI to ensure it integrates well.

📤 **Outputs**
- [x] Do not send back as outputs the same infos you already have in your properties.
- [x] If you do not have any output use `VoidOutput`.
- [x] Do not output twice the same infos (eg: a status code, an error code saying the same thing...).

santhoshh-kumar committed:
feat: implement Microsoft 365 OneShare subplugin

Features:
- Upload: Support for small and large file uploads with resumable sessions
- Download: Download files from OneDrive/SharePoint to internal storage
- Create: Create files and folders with content support
- Delete: Remove files and folders
- Export: Export files to different formats (PDF, HTML, etc.)
- ListFiles: List and browse drive contents with pagination
- Trigger: React to new file creation, existing file updates or both using Delta API

Implementation details:
- Microsoft Graph API integration with Azure AD authentication
- Support for both client secret and certificate authentication
- Abstract base classes for connection management
- Comprehensive test coverage with integration tests
- Icon resources for UI integration

Dependencies:
- Microsoft Graph SDK v6.54.0
- Azure Identity library for authentication
- Jackson for JSON serialization

santhoshh-kumar committed:
Merge branch 'main' into feat/oneshare

santhoshh-kumar committed:
Refactor(tests): Add mocking for tests and skip tests that require credentials.

- Updated DeleteTest to mock GraphServiceClient and verify delete operations.
- Enhanced DownloadTest with mocks for file download functionality and added tests for different IDs.
- Refactored ExportTest to mock PDF export functionality and validate output.
- Improved ListFilesTest to mock file listing and verify returned file details.
- Updated TriggerTest to mock delta requests and validate trigger execution without flooding.
- Enhanced UploadTest to mock file uploads and verify behavior with conflict handling.

santhoshkumarCodes commented on .github/workflows/main.yml:
```diff
@@ -5,13 +5,15 @@ on:
     - cron: '0 4 * * 1,2,3,4,5'
   push:
     branches:
+      - master
```
Sorry! I copy pasted it from gcp plugin. Will remove it now

Malaydewangan09 commented on src/main/java/io/kestra/plugin/microsoft365/oneshare/AbstractOneShareTask.java:
```diff
@@ -0,0 +1,16 @@
+@EqualsAndHashCode(callSuper = true)
+@Getter
+@NoArgsConstructor
+public abstract class AbstractOneShareTask extends AbstractGraphConnection {
```
Why this is empty? Don't we have anything that is used in all tasks and we could extract them?

santhoshh-kumar committed:
CHORE: Remove master from `main.yml` since we are not using it

Malaydewangan09 commented on build.gradle:
```diff
@@ -137,6 +157,9 @@ test {
 
 jacocoTestReport {
     dependsOn test
+    reports {
```
Is this taken from any plugin?

Malaydewangan09 commented on build.gradle:
```diff
@@ -188,6 +211,7 @@ jar {
 }
 
 shadowJar {
+    zip64 true
```
Same comment as above

santhoshkumarCodes replied:
Yes both of them added from `gcp plugin`. Should I remove both of them...?

Malaydewangan09 reviewed (commented):
@santhoshkumarCodes, LGTM
Would be great, if you could add IT tests and guard them with `@EnabledIf/DisabledIf`
WDYT?

santhoshkumarCodes commented on build.gradle:
```diff
@@ -45,14 +45,30 @@ dependencies {
 
-    // kestra
+    // Micronaut
+    compileOnly "io.micronaut:micronaut-http-client"
```
Sorry! This was added  when I was trying to implement webhook based approach using `subscription api` which requires immediate `ACK` after receiving response from `graph api`. Will remove it now.

santhoshkumarCodes commented:
> guard them with `@EnabledIf/DisabledIf` 

Yes, I think it would be better!

santhoshh-kumar committed:
feat(oneshare): Refactor OneShare tasks and improve integration tests

- Move driveId to AbstractOneShareTask to avoid duplication.

- Use @EnabledIf to conditionally run integration tests based on credential availability.

- Remove unused Micronaut dependencies from build.gradle.

santhoshkumarCodes commented:
@Malaydewangan09 - I've added `@EnabledIf/DisabledIf` guard for the IT tests and also addressed all the review comments.

Malaydewangan09 commented on src/main/java/io/kestra/plugin/microsoft365/oneshare/Export.java:
```diff
@@ -0,0 +1,116 @@
+        description = "Supported formats include pdf, glb, html, jpg, and png."
+    )
+    @NotNull
+    private Property<String> format;
```
If these are exhaustive, could we use Enum here?

Malaydewangan09 commented on src/main/java/io/kestra/plugin/microsoft365/oneshare/ListFiles.java:
```diff
@@ -0,0 +1,138 @@
+        )
+    }
+)
+public class ListFiles extends AbstractOneShareTask implements RunnableTask<ListFiles.Output> {
```
No issues, you can keep it as `List` only

Malaydewangan09 replied:
@santhoshkumarCodes, I think you forgot this?

santhoshh-kumar committed:
feat: Add file count to ListFiles output and set default parentId in Upload

santhoshh-kumar committed:
feat: Update Export task to use enum for format and improve test cases

bishalbera commented:
> Alright, thanks @bishalbera! The example flows for OA are included in the PR description - they cover all tasks and trigger. Take your time to finish your current PR, and then you can come back to this one.

@santhoshkumarCodes Just a doubt — did you get any confirmation email after registering for the Hackfest organized by Kunal? I didn’t get one :(

bishalbera commented:
<img width="1919" height="926" alt="image" src="https://github.com/user-attachments/assets/391480a1-e266-4e7e-8688-233ec42501f7" />

<img width="1916" height="804" alt="image" src="https://github.com/user-attachments/assets/bd7710de-6272-4f97-9d03-ca3e5e6ab790" />

<img width="1906" height="795" alt="image" src="https://github.com/user-attachments/assets/2ffca794-b2c7-44ae-a83f-4d2294ee10e9" />

<img width="1907" height="513" alt="image" src="https://github.com/user-attachments/assets/95288798-58c9-41e5-b061-fe39e10bfde0" />

santhoshkumarCodes commented:
Thanks! @bishalbera - I will look into it

santhoshh-kumar committed:
feat: Enhance error handling and input validation in OneDrive/SharePoint tasks

- Added input validation for item ID and format in Export task.
- Improved error handling for various API exceptions in Export and ListFiles tasks.
- Implemented input validation for path in Trigger task.
- Enhanced error handling during delta queries in Trigger task.
- Added validation for file name and slice size in Upload task.
- Introduced new ItemType enum to represent item types for creation tasks.
- Updated tests to use the new ItemType enum for folder creation.

santhoshkumarCodes commented:
@bishalbera - I have updated the flow files to overcome path error for `Trigger` and format error for `export` task. Can you try running all 3 flows in the `PR Description` when you got time?

santhoshkumarCodes commented:
Thanks! I'll check @bishalbera

santhoshkumarCodes commented:
@bishalbera - I have a doubt regarding `Trigger`. Have you uploaded the file in the `/` (root) path as we have set the path to monitor as `/` .Can you please check that for me?

santhoshh-kumar committed:
feat: Add detailed logging for delta link and state management in Trigger

santhoshkumarCodes commented:
Regarding `Trigger`, We should have received the deltaLink for next execution to trigger updates. But the logs shows that 

`No delta link found` - This seems like a state management issue or maybe the graph is not at all sending the `deltaLink`. So I've added full logging in my latest commit to debug this issue.

santhoshkumarCodes commented:
@bishalbera - Can you please run and test the trigger in the `/` (root) folder with my latest commit and share the logs.

santhoshkumarCodes commented:
I think we don't need to remove any of the `debug` logging added in the last commit as it anyways gonna help users. WDYT `maintainers` ? 

And one more thing, the error from the [screenshot](https://github.com/kestra-io/plugin-microsoft365/pull/5#issuecomment-3476383262) is due to using `dotted` notation instead of `bracketed` notation for `Download task itemId`. So, there is no issue, am I right?

santhoshh-kumar committed:
feat: Update README and package-info to clarify OneShare plugin features and SharePoint integration

Malaydewangan09 commented on src/main/java/io/kestra/plugin/microsoft365/oneshare/Create.java:
```diff
@@ -0,0 +1,257 @@
+                    try {
+                        result = client.drives().byDriveId(rDriveId).items().byDriveItemId(rParentId).children().post(driveItem);
+                    } catch (ApiException e) {
+                        if (e.getResponseStatusCode() == 404) {
```
@santhoshkumarCodes, these exception handling is quite repetitive, could we do something for this? WDYT?

santhoshkumarCodes replied:
I've thought that too...  But each of the response code in each tasks have different strings to be logged to the user. Even if we make a common util file - we need to handle it using switch cases by having a seperate `util` class which makes the util file itself complex. Do you have any other ideas..?

Malaydewangan09 commented on src/main/java/io/kestra/plugin/microsoft365/oneshare/Download.java:
```diff
@@ -0,0 +1,214 @@
+                tempFile = runContext.workingDir().createTempFile().toFile();
+                fos = new FileOutputStream(tempFile);
+                
+                byte[] buffer = new byte[8192];
```
Couldn't we use something like `inputStream.transferTo(outputStream)`, to avoid this manual loop and managing streams?
cc @fdelbrayelle

santhoshh-kumar committed:
feat: Refactor Download task to use try-with-resources and transferTo method for InputStream and FileOutputStream

santhoshh-kumar committed:
feat: Rename `ListFiles` to `List`  and update references.

Malaydewangan09 reviewed (approved):
LGTM, let's wait for @fdelbrayelle final review before we merge 🚀

santhoshh-kumar committed:
Merge branch 'main' into feat/oneshare

santhoshh-kumar commented:
Sorry didn't see it. Resolved now! @fdelbrayelle

Malaydewangan09 committed:
Merge branch 'main' into feat/oneshare

santhoshh-kumar committed:
Merge branch 'main' into feat/oneshare

fdelbrayelle commented:
Hello @santhoshh-kumar :wave: I know but we have 30+ PRs to review, our own internal highlights, 40+ notifications on GitHub... So we do what we can do. Be patient, this will be merged for sure.

santhoshh-kumar commented:
Ohh! I'm sorry I just want to remind you. I thought maybe you've forgotten.
